File: python/composio/utils/safe_path.py
"""Path containment primitives for untrusted input.

Every field of an API response is untrusted, including slugs, identifiers, and
filenames. A compromised toolkit or a MITM on the Composio API can set them to
anything. When such a value becomes part of a filesystem path, it must pass
through this module.

Two rules make the difference between a real check and a decorative one:

1. **Anchor on a constant.** Containment is only meaningful when the root is a
   trusted value. Checking a derived path against a directory that untrusted
   input helped build compares tainted against tainted and always passes. A
   directory built as ``root / untrusted_a / untrusted_b`` and then validated
   against that same built directory is checked against a reference the input
   was free to move.

2. **Validate before touching the filesystem.** No ``mkdir``, no ``open``, until
   the final path is known to be inside the root. Otherwise a rejected write
   still leaves attacker-chosen directories behind.

:func:`secure_join` enforces both. Prefer it over hand-rolled
``resolve() / is_relative_to()`` pairs — see ``python/AGENTS.md``.
"""

from __future__ import annotations

import os
import re
import sys
import typing as t
from pathlib import Path, PureWindowsPath

from composio.exceptions import UnsafePathComponentError

SAFE_COMPONENT_REGEX = re.compile(r"^[A-Za-z0-9_-]+$")
"""Characters permitted in a path component derived from untrusted input.

Deliberately identical to ``SLUG_REGEX`` in ``custom_tool_types.py``, which the
SDK already enforces on client-created custom tools. Backend-fetched tools were
never held to the same standard; they are now.

Every one of the 1070 toolkit slugs in the catalog snapshot at
``ts/packages/cli/src/generated/toolkit-slugs.ts`` conforms, so this is not
expected to reject anything legitimate. That snapshot is not the live catalog:
a production slug containing a dot or a non-ASCII character would hard-fail its
downloads rather than degrade.
"""

MAX_COMPONENT_LENGTH = 128
"""Upper bound on a single component, well under the 255-byte limit common to
ext4/APFS/NTFS. Keeps a long slug from raising ``OSError`` mid-write."""

WINDOWS_RESERVED_NAMES = frozenset(
    {"CON", "PRN", "AUX", "NUL"}
    | {f"COM{i}" for i in range(1, 10)}
    | {f"LPT{i}" for i in range(1, 10)}
    | {f"COM{i}" for i in "¹²³"}
    | {f"LPT{i}" for i in "¹²³"}
)
"""Reserved DOS device names. Writing to one on Windows targets the device
rather than a file. Rejected on every platform so behaviour does not diverge
between a POSIX developer machine and a Windows deployment."""


def is_inside_dir(child: Path, parent: Path) -> bool:
    """True iff ``child`` equals ``parent`` or is nested inside it on a path
    component boundary.

    ``/tmp/foo`` contains ``/tmp/foo/bar`` but NOT ``/tmp/foo-bar`` — a plain
    string prefix test gets that wrong. Comparison is case-insensitive on
    Windows, where ``C:\\Foo`` and ``c:\\foo`` are the same directory.

    Both arguments are assumed absolute and normalized.
    """
    try:
        child_str = str(child)
        parent_str = str(parent)
        if sys.platform == "win32":
            child_str = child_str.lower()
            parent_str = parent_str.lower()
        if child_str == parent_str:
            return True
        sep = os.sep
        parent_with_sep = parent_str if parent_str.endswith(sep) else parent_str + sep
        return child_str.startswith(parent_with_sep)
    except OSError:
        return False


def assert_safe_path_component(value: str, *, label: str = "path component") -> str:
    """Return ``value`` unchanged if it is safe to use as a single path
    component, else raise.

    Fails closed. Rejects traversal (``..``), separators of either platform,
    absolute paths, drive letters, NUL bytes, reserved device names, and
    anything outside :data:`SAFE_COMPONENT_REGEX`.

    :raises UnsafePathComponentError: when ``value`` is unsafe.
    """
    if not isinstance(value, str) or not value:
        raise UnsafePathComponentError(
            f"Refusing to build a path from an empty or non-string {label}: {value!r}"
        )

    # `PureWindowsPath` treats both `/` and `\` as separators, so a single check
    # catches `../x` and `..\x` regardless of the host platform. A slug crafted
    # for a Windows target must not slip through on a POSIX build machine.
    as_windows_path = PureWindowsPath(value)
    if len(as_windows_path.parts) != 1 or as_windows_path.anchor:
        raise UnsafePathComponentError(
            f"Refusing to build a path from a {label} containing path separators "
            f"or a drive letter: {value!r}"
        )

    if len(value) > MAX_COMPONENT_LENGTH:
        raise UnsafePathComponentError(
            f"Refusing to build a path from a {label} longer than "
            f"{MAX_COMPONENT_LENGTH} characters: {value[:32]!r}... "
            f"({len(value)} characters)"
        )

    # `.` and `..` are excluded by the regex (no `.` in the character class),
    # as are NUL bytes and every separator. The explicit checks above exist to
    # produce a precise error message rather than a generic pattern mismatch.
    #
    # `fullmatch`, not `match`: in a `match`, `$` also matches just before a
    # single trailing newline, so `"GMAIL\n"` would satisfy `^[A-Za-z0-9_-]+$`
    # and reach the filesystem with a control character in the name.
    if not SAFE_COMPONENT_REGEX.fullmatch(value):
        raise UnsafePathComponentError(
            f"Refusing to build a path from an unsafe {label}: {value!r}. "
            f"Expected only letters, digits, underscores, and hyphens "
            f"(pattern {SAFE_COMPONENT_REGEX.pattern})."
        )

    if value.upper() in WINDOWS_RESERVED_NAMES:
        raise UnsafePathComponentError(
            f"Refusing to build a path from a reserved device name as {label}: {value!r}"
        )

    return value


def safe_basename(name: str, *, label: str = "filename") -> str:
    """Collapse an untrusted filename to a bare, writable basename.

    Filenames need their own validator: :func:`assert_safe_path_component`
    forbids ``.``, which nearly every real filename contains. This applies the
    remaining checks — no separators, no traversal, no NUL, bounded length, no
    reserved device name — to the one component a server most directly controls.

    ``PureWindowsPath`` treats both ``/`` and ``\\`` as separators, so a name
    crafted for a Windows target (``..\\..\\evil``) is stripped even when the
    SDK runs on POSIX, where ``Path(...).name`` would return it intact.

    Names that leave no usable basename are refused rather than replaced with a
    generated one: a response that cannot name its own file is malformed or
    hostile, and inventing a name would hide that. ``.`` and the empty string
    both basename to ``""``, which makes an output path equal to its own
    directory and surfaces as ``IsADirectoryError`` at write time.

    :raises UnsafePathComponentError: when ``name`` yields no usable basename or
        is unsafe to write.
    """
    if not isinstance(name, str):
        raise UnsafePathComponentError(
            f"Refusing to write a non-string {label}: {name!r}"
        )

    raw_basename = PureWindowsPath(name).name
    if not raw_basename or not raw_basename.strip() or set(raw_basename) == {"."}:
        raise UnsafePathComponentError(
            f"Path traversal detected: {label} {name!r} leaves no usable "
            "basename to write to."
        )
    if "\x00" in raw_basename:
        raise UnsafePathComponentError(
            f"Refusing to write {label} containing a NUL byte: {name!r}"
        )
    if any(ord(char) < 32 or char in '<>:"|?*' for char in raw_basename):
        raise UnsafePathComponentError(
            f"Refusing to write {label} containing characters reserved by "
            f"Windows: {name!r}"
        )
    if raw_basename.endswith((" ", ".")):
        raise UnsafePathComponentError(
            f"Refusing to write {label} ending in a space or dot: {name!r}"
        )

    basename = raw_basename.strip()
    try:
        encoded_length = len(os.fsencode(basename))
    except UnicodeEncodeError as e:
        raise UnsafePathComponentError(
            f"Refusing to write {label} containing invalid Unicode: {name!r}"
        ) from e
    if encoded_length > MAX_COMPONENT_LENGTH:
        raise UnsafePathComponentError(
            f"Refusing to write {label} longer than {MAX_COMPONENT_LENGTH} bytes: "
            f"{basename[:32]!r}... ({encoded_length} bytes)"
        )
    # Compare everything before the first dot: on Windows `NUL.tar.gz` opens
    # the null device just as `NUL` does, so any number of extensions provides
    # no protection.
    device_name = basename.split(".", 1)[0].rstrip(" ").upper()
    if device_name in WINDOWS_RESERVED_NAMES:
        raise UnsafePathComponentError(
            f"Refusing to write {label} that is a reserved device name: {name!r}"
        )
    return basename


def resolve_root(root: t.Union[str, Path]) -> Path:
    """Normalize a trusted root to an absolute, symlink-resolved path.

    Every containment check must derive its anchor through this one function.
    When two call sites normalize differently — one expanding ``~`` and one not
    — the check compares mismatched paths and rejects legitimate writes while
    reporting them as attacks. Sharing the normalization is what keeps the two
    ends of a containment check comparable.
    """
    expanded = Path(root).expanduser()
    try:
        return expanded.resolve(strict=False)
    except OSError:
        return expanded


def secure_basename_join(
    base: t.Union[str, Path],
    name: str,
    *,
    root: t.Optional[t.Union[str, Path]] = None,
    label: str = "filename",
) -> Path:
    """Join a single untrusted filename under ``base``, contained within ``root``.

    The filename counterpart to :func:`secure_join`, which cannot be reused here
    because it forbids ``.`` — correct for a slug, wrong for ``report.pdf``.

    ``root`` defaults to ``base`` but is separate for the download path, where
    ``base`` is a per-tool subdirectory that untrusted slugs helped build and
    only the configured ``root`` above it is trusted. Anchoring on ``base``
    there would check the result against a directory those slugs had moved.

    :raises UnsafePathComponentError: when ``name`` is unsafe, or when the
        result escapes ``root``.
    """
    resolved_base = resolve_root(base)
    resolved_root = resolved_base if root is None else resolve_root(root)
    candidate = resolved_base / safe_basename(name, label=label)
    if not is_inside_dir(candidate.resolve(), resolved_root):
        raise UnsafePathComponentError(
            f"Path traversal detected: {label} {name!r} resolves to "
            f"{candidate.resolve()}, which is outside {resolved_root}."
        )
    return candidate


def secure_join(root: t.Union[str, Path], *components: str) -> Path:
    """Join untrusted ``components`` beneath the trusted ``root``.

    ``root`` is the sole anchor of trust and must not itself be derived from
    untrusted input — that is the whole point. Each component is validated by
    :func:`assert_safe_path_component`, then the joined result is resolved and
    re-checked against the resolved root. The second check is belt-and-braces:
    it catches a symlink inside ``root`` pointing outside it, which per-component
    validation cannot see.

    Performs no filesystem writes; the caller creates directories only after
    this returns.

    :raises UnsafePathComponentError: when any component is unsafe, or when the
        joined path escapes ``root``.
    """
    resolved_root = resolve_root(root)

    safe_components = [
        assert_safe_path_component(component) for component in components
    ]

    candidate = resolved_root.joinpath(*safe_components)
    try:
        resolved_candidate = candidate.resolve(strict=False)
    except OSError:
        resolved_candidate = candidate

    if not is_inside_dir(resolved_candidate, resolved_root):
        raise UnsafePathComponentError(
            f"Refusing to write outside the configured directory: "
            f"{components!r} resolves to {resolved_candidate}, "
            f"which is outside {resolved_root}."
        )

    return resolved_candidate

File: ts/packages/core/src/utils/safePath.ts
/**
 * Filename containment for untrusted input.
 *
 * Every field of an API response is untrusted, including filenames. When such a
 * value becomes part of a filesystem path it must pass through this module
 * first, and validation must happen before any `mkdir` or write so a rejected
 * value leaves nothing behind on disk.
 *
 * TypeScript counterpart of `safe_basename` in `python/composio/utils/safe_path.py`.
 * The two use the same validation order and whitespace rules. Path extraction
 * is intentionally limited here; see `untrustedBasename`.
 *
 * Pure string logic with no filesystem access: no static `node:*` imports, so
 * it is usable from edge/workerd builds.
 */
import { ValidationError } from '../errors/ValidationErrors';

/**
 * Upper bound on a filename, well under the 255-byte limit common to
 * ext4/APFS/NTFS. Keeps a long server-supplied name from failing with a raw
 * `ENAMETOOLONG` mid-write. Matches `MAX_COMPONENT_LENGTH` in the Python SDK.
 */
export const MAX_FILENAME_BYTES = 128;

/**
 * Reserved DOS device names. Writing to one on Windows targets the device
 * rather than a file. Rejected on every platform so behavior does not diverge
 * between a POSIX developer machine and a Windows deployment.
 */
const WINDOWS_RESERVED_NAMES: ReadonlySet<string> = new Set([
  'CON',
  'PRN',
  'AUX',
  'NUL',
  ...Array.from({ length: 9 }, (_, i) => `COM${i + 1}`),
  ...Array.from({ length: 9 }, (_, i) => `LPT${i + 1}`),
  ...['¹', '²', '³'].flatMap(superscript => [`COM${superscript}`, `LPT${superscript}`]),
]);

/** Control characters and characters reserved by Windows. */
const WINDOWS_INVALID_CHARS = /[\u0000-\u001f<>:"|?*]/;

/**
 * Python str.strip() whitespace: Unicode White_Space plus U+001C–U+001F.
 * Unlike JavaScript trim(), this includes U+0085 and preserves U+FEFF.
 */
const PYTHON_WHITESPACE =
  '[\\u0009-\\u000d\\u001c-\\u0020\\u0085\\u00a0\\u1680\\u2000-\\u200a\\u2028\\u2029\\u202f\\u205f\\u3000]';
const SURROUNDING_WHITESPACE = new RegExp(`^${PYTHON_WHITESPACE}+|${PYTHON_WHITESPACE}+$`, 'g');

/**
 * Returns the last path segment of `value`, treating both `/` and `\` as
 * separators, or `''` when there is none.
 *
 * `path.basename()` on POSIX only splits on `/`, so a name crafted for a
 * Windows target (`..\..\evil`) comes back intact on macOS and Linux. Splitting
 * on both and removing an initial drive prefix also handles `C:report.txt`.
 * This is not a full `PureWindowsPath` parser: dot components stay literal,
 * and UNC anchors are not interpreted.
 *
 * Never throws: this is a display value as well as the input to
 * {@link safeBasename}, and it is read while constructing download errors.
 */
export function untrustedBasename(value: string): string {
  const segments = value.replace(/^[a-z]:/i, '').split(/[/\\]+/);
  // Trailing separators name the same file: `a/b/` and `a/b` both basename to `b`.
  while (segments.length > 0 && segments[segments.length - 1] === '') {
    segments.pop();
  }
  return segments[segments.length - 1] ?? '';
}

/**
 * True when `value` contains a surrogate code unit that is not part of a valid
 * pair. Such a string cannot be encoded as UTF-8; `TextEncoder` silently
 * substitutes U+FFFD, which would write a file under a name the response never
 * specified. `String.prototype.isWellFormed()` is ES2024 and this package
 * targets es2022, hence the manual scan.
 */
function hasLoneSurrogate(value: string): boolean {
  for (let i = 0; i < value.length; i++) {
    const code = value.charCodeAt(i);
    if (code >= 0xd800 && code <= 0xdbff) {
      const next = value.charCodeAt(i + 1);
      if (!(next >= 0xdc00 && next <= 0xdfff)) {
        return true;
      }
      i++;
    } else if (code >= 0xdc00 && code <= 0xdfff) {
      return true;
    }
  }
  return false;
}

/**
 * Collapses an untrusted filename to a bare, writable basename.
 *
 * Applies, in order: no usable basename (empty or a run of dots once
 * surrounding whitespace is stripped), NUL bytes, control and Windows-reserved
 * characters, trailing space or dot, invalid Unicode, a byte-length bound, and
 * reserved device names. This is the order `safe_basename` uses in the Python
 * SDK. This implementation rejects lone surrogates rather than relying on
 * Python filesystem encoding behavior.
 *
 * Names that leave no usable basename are refused rather than replaced with a
 * generated one: a response that cannot name its own file is malformed or
 * hostile, and inventing a name would hide that. `""`, `"."` and `".."` make
 * an output path equal to, or escape, its own directory, which surfaces as a
 * raw `EISDIR` at write time instead of a validation error.
 *
 * The usability check runs on the *trimmed* basename because that is what gets
 * written: Python-compatible stripping removes whitespace, so `"\u00a0.\u00a0"` would
 * otherwise pass a check on the raw segment and then be written as `"."`. The
 * hazard checks that follow run on the raw segment so a trailing ASCII space or
 * dot is refused, not trimmed away. Check the stripped value for trailing dots
 * too, since stripping whitespace can expose one.
 *
 * @param name - The untrusted filename or relative path to reduce.
 * @param label - How the value is described in error messages.
 * @returns The validated bare basename, with surrounding whitespace trimmed.
 * @throws ValidationError if `name` yields no usable basename or is unsafe to write.
 */
export function safeBasename(name: string, label: string = 'filename'): string {
  const rawBasename = untrustedBasename(name);
  const basename = rawBasename.replace(SURROUNDING_WHITESPACE, '');

  if (!basename || /^\.+$/.test(basename)) {
    throw new ValidationError(
      `Path traversal detected: ${label} ${JSON.stringify(name)} leaves no usable basename to write to.`
    );
  }
  if (rawBasename.includes('\u0000')) {
    throw new ValidationError(
      `Refusing to write ${label} containing a NUL byte: ${JSON.stringify(name)}`
    );
  }
  if (WINDOWS_INVALID_CHARS.test(rawBasename)) {
    throw new ValidationError(
      `Refusing to write ${label} containing characters reserved by Windows: ${JSON.stringify(name)}`
    );
  }
  if (rawBasename.endsWith(' ') || rawBasename.endsWith('.') || basename.endsWith('.')) {
    throw new ValidationError(
      `Refusing to write ${label} ending in a space or dot: ${JSON.stringify(name)}`
    );
  }
  if (hasLoneSurrogate(basename)) {
    throw new ValidationError(
      `Refusing to write ${label} containing invalid Unicode: ${JSON.stringify(name)}`
    );
  }

  const encodedLength = new TextEncoder().encode(basename).length;
  if (encodedLength > MAX_FILENAME_BYTES) {
    throw new ValidationError(
      `Refusing to write ${label} longer than ${MAX_FILENAME_BYTES} bytes: ` +
        `${JSON.stringify(basename.slice(0, 32))}... (${encodedLength} bytes)`
    );
  }

  // Compare everything before the first dot: on Windows `NUL.tar.gz` opens the
  // null device just as `NUL` does, so extensions provide no protection.
  const deviceName = basename.split('.', 1)[0].replace(/ +$/, '').toUpperCase();
  if (WINDOWS_RESERVED_NAMES.has(deviceName)) {
    throw new ValidationError(
      `Refusing to write ${label} that is a reserved device name: ${JSON.stringify(name)}`
    );
  }

  return basename;
}
